draft/XXXX-configuration_handling/README.md

```diff
@@ -0,0 +1,202 @@
+---
+RFC: XXXX
+Author: Roni Choudhury
+Created: 2020-06-29
+Created: 2020-12-09
+Last-Modified: 2020-12-09
+---
+
+# RFC XXXX: Configuration Handling
+
+One of the core features of our vision for Tinker Engine is the ability to
+accept a kind of "meta-configuration" from which concrete configurations are
+generated, used to drive individual protocol runs, with results aggregated
+together in some way that enables post-hoc data analysis. This RFC lays out some
+ideas for some of the core features such a handler will have, along with an
+overall architecture for configuration hanlding in Tinker Engine.
+
+## Proposals
+
+This RFC will discuss a few concrete proposals. One is the overall architcture
+by which the Tinker Engine runtime receives, processes, and submits
+configuration data to a protocol. Another is the method by which a configuration
+file can express variation in its parameters. A third is a fallback mechanism
+allowing programmatic configuration handling when the declarative methods are
+not sufficient for some use case.
+
+### Overall Architecture
+
+Currently, invoking the `tinker` CLI requires specifying a configuration file,
+but currently Tinker does nothing with this file. The new behavior will be for
+Tinker to load the specified file and parse it as YAML (bearing in mind that
+valid JSON is also valid YAML). If the file cannot be opened, or if the contents
+do not parse properly, Tinker will exit with an error condition.
+
+The YAML content describes a "meta-configuration" (see below) that requires
+processing to turn it into a sequence of configurations, one per experimental
+run. This sequence of configurations will be used to construct one protocol per
+configuration, running each to complete one trial of the experimental protocol.
+
+This will be accomplished by sending each configuration dictionary to the
+protocol initializer (so that client logic can dictate how the configuration is
+used to set the protocol up), and then also as an argument to the protocol's
+`run()` method (allowing client logic to make use of the configuration data to
+direct the protocol runtime).
+
+The design of the particular experiment will be captured in the variation of
+configuration values, as expressed via *declarative meta-configuration* or
+*programmatic configuration*.
+
+### Declarative Meta-Configuration
+
+In many cases, the user of Tinker Engine will want to run a protocol multiple
+times, changing some set of parameters in order to observe the effect those
+changes have on the results of the run. In this way, Tinker Engine supports
+computational experiments through the use of the scientific method.
```

Comment:
I would suggest curly brackets with comma value separators  {1, 2, 3}.  That way we can distinguish vectors [1, 2, 3] parameters and dictionary parameters {'a': 1, 'b': 2}.

Reply:
I don't think the curly bracket syntax is valid YAML, but part of the work being proposed here is to figure out the right kind of schema for expressing exactly this type of distinction.

In the examples I made up, I signal the difference using the `values` key, whose argument is expected to be an array of values to iterate through. We'll need to lock down what the full set of these signal words are (and we don't need to do so immediately; we can incrementally add new ones and make adjustments as we go).

Reply:
How would we identify a parameter which is a vector of values then?  E.g. https://gitlab.kitware.com/darpa_learn/learn/-/blob/master/learn/algorithms/HPT/config.py#L100

Reply:
I think it would be the same as anything else. If you had something like

```yaml
foo: [1, 2, 3]
```

that would mean "the value of the parameter `foo` is the array `[1, 2, 3]`", but by contrast

```yaml
foo:
  values: [1, 2, 3]
```

would mean "the value of the parameter `foo` is each of the values `1`, `2`, and `3` during three separate runs", and

```yaml
foo:
  values: [[1, 2, 3], [4, 5, 6]]
```

would mean "the value of the parameter `foo` is each of the arrays `[1, 2, 3]` and `[4, 5, 6]` during two separate runs".

We might want a better signal word than `values`; maybe `iterate` would be good. But the distinction is the use of a signal word of some sort to indicate "splitting the universe" into separate runs using the sequence of values.

Does that make sense?

Reply:
How many protected names (like 'values') will we define here?  Maybe we should have a clearly defined list

Reply:
Well, the idea is that as we define features that the system can take advantage of, we'd track these signal words in documentation, and in the JSON schema we're discussing in the other thread. To answer your question directly, in this initial round, maybe we'd just define the one (`values` or `iterate`). I agree with you that we need to define them explicitly, but we don't need to do all of the possible ones right this minute.

Reply:
I'm fine with this, I just think we need to be clean on what names are protected in the tinker language.

Reply:
This would be necessary to do in the documentation.  Might be nice here but more might pop up.  Resolving for now.